README
This gem/Rails plugin contains common code for powering iPhone applications. It
is particularly suitable as a server-side partner to the ZergSupport toolkit,
also available on Github under the MIT license at
http://github.com/costan/zergsupport

 

CHANGELOG
v0.0.9. Added support for device attributes stored in Tempfile instances.

v0.0.8. Added flushing to daemon-optimized push notification API.

v0.0.7. New API for push notifications, better suited for daemons.

v0.0.6. New push certificates read method for Ruby 1.8.6 on Heroku.

v0.0.5. Renamed device_token to push_token throughout the API.

v0.0.4. Updated CryptoSupport with the push token device attribute.  

v0.0.3. Push Notification support.

v0.0.2. CryptoSupport finger-prints computation.

v0.0.1. Initial release. In-App Purchase receipt validation.


test/validate_receipt_test.rb
# Author:: Victor Costan
# Copyright:: Copyright (C) 2009 Zergling.Net
# License:: MIT

require 'imobile'

require 'time'
require 'test/unit'


class ValidateReceiptTest < Test::Unit::TestCase
  def setup
    super
    
    testdata_path = File.join(File.dirname(__FILE__), '..', 'testdata')    
    @forged_sandbox_blob = File.read File.join(testdata_path,
                                               'forged_sandbox_receipt')
    @valid_sandbox_blob = File.read File.join(testdata_path,
                                              'valid_sandbox_receipt')
  end
  
  def test_valid_sandbox_receipt
    golden_date = Time.parse('2009-07-22 18:52:46 EEST')
    
    receipt = Imobile.validate_receipt @valid_sandbox_blob, :sandbox
    assert receipt, "Genuine receipt failed validation"
    
    assert_equal 1, receipt[:quantity], 'Wrong quantity'
    assert_equal 'us.costan.ZergSupportTests', receipt[:bid], 'Wrong bundle ID'
    assert_equal '1.9.8.3', receipt[:bvrs], 'Wrong bundle version'
    assert_equal 'net.zergling.ZergSupport.sub_cheap', receipt[:product_id],
                 'Wrong product ID'
    assert_equal '324740515', receipt[:item_id], 'Wrong item (iTunes app) ID' 
    assert_equal '1000000000016600', receipt[:transaction_id],
                 'Wrong transaction ID' 
    assert_equal '1000000000016600', receipt[:original_transaction_id],
                 'Wrong original transaction ID' 
    assert_equal golden_date.to_f,
                 Time.parse(receipt[:purchase_date].to_s).to_f,
                 'Wrong purchase date'    
    assert_equal golden_date.to_f,
                 Time.parse(receipt[:original_purchase_date].to_s).to_f,
                 'Wrong original purchase date'
  end
  
  def test_forged_sandbox_receipt
    assert_equal false,
                 Imobile.validate_receipt(@forged_sandbox_blob, :sandbox),
                 "Forged receipt passed validation"
  end
  
  # TODO(costan): add tests against the real servers, as soon as someone donates
  #               a receipt
end


test/crypto_app_fprint_test.rb
# Author:: Victor Costan
# Copyright:: Copyright (C) 2009 Zergling.Net
# License:: MIT

require 'imobile'

require 'time'
require 'test/unit'

require 'rubygems'
require 'flexmock/test_unit'


class CryptoAppFprintTest < Test::Unit::TestCase
  def setup
    super
    
    testdata_path = File.join(File.dirname(__FILE__), '..', 'testdata')
    @device_attrs = File.open(File.join(testdata_path,
                                        'device_attributes.yml')) do |f| 
      YAML.load f
    end
    
    @mock_binary = '1a2b3c4d5e6f7g8h' * 16384
  end
  
  def test_device_fprint
    fprint = Imobile::CryptoSupportAppFprint.hex_device_fprint @device_attrs
    assert_equal '3231b35f8466d60a6ae00122d2530e65', fprint
  end
  
  class MockTempfile
    def initialize(string)
      @string = string
    end
    def read
      @string
    end
  end
  
  def test_device_fprint_from_tempfile
    attrs = {} 
    @device_attrs.each { |k, v| attrs[k] = MockTempfile.new v }
    fprint = Imobile::CryptoSupportAppFprint.hex_device_fprint @device_attrs
    assert_equal '3231b35f8466d60a6ae00122d2530e65', fprint
  end

  def test_crypto_app_fprint
    mock_binary_path = '/binary/path'
    flexmock(File).should_receive(:read).with(mock_binary_path).
                   and_return(@mock_binary)
    fprint = Imobile.crypto_app_fprint @device_attrs, mock_binary_path
    gold_fprint =
        'd84630797d0b19d7d8403705bcc154cfbd82ff386587875193f7d99f01348014' 
    assert_equal gold_fprint, fprint
  end
end


test/push_notification_test.rb
# Author:: Victor Costan
# Copyright:: Copyright (C) 2009 Zergling.Net
# License:: MIT

require 'imobile'

require 'time'
require 'test/unit'

require 'rubygems'
require 'flexmock/test_unit'


class PushNotificationTest < Test::Unit::TestCase
  def setup
    super
    
    testdata_path = File.join(File.dirname(__FILE__), '..', 'testdata')
    @dev_cert_path = File.join(testdata_path, 'apns_developer.p12') 
    @prod_cert_path = File.join(testdata_path, 'apns_production.p12')
    
    @hex_dev_token = File.read File.join(testdata_path, 'sandbox_push_token')
    @dev_token = File.read File.join(testdata_path, 'sandbox_push_token.bin')
    
    @notification = {:aps => {:alert => 'imobile test notification'}}
    @encoded_notification = File.read File.join(testdata_path,
                                                'encoded_notification')
  end
  
  def test_read_push_certificate
    cert_data = Imobile::PushNotifications.read_certificate @dev_cert_path
    assert cert_data, "Dev certificate didn't load"    
    assert_equal :sandbox, cert_data[:server_type],
                 "Dev certificate mistaken for prod"
    assert cert_data[:key].kind_of?(OpenSSL::PKey::PKey),
           "Dev certificate does not contain a key"
    assert(/Q686F7Z6YU/ =~ cert_data[:certificate].subject.to_s,
           "Wrong data in dev certificate #{cert_data[:certificate].inspect}")

    prod_blob = File.read @prod_cert_path
    cert_data = Imobile::PushNotifications.read_certificate prod_blob 
    assert cert_data, "Prod certificate didn't load"    
    assert_equal :production, cert_data[:server_type],
                 "Prod certificate mistaken for dev"
    assert cert_data[:key].kind_of?(OpenSSL::PKey::PKey),
           "Prod certificate does not contain a key"
    assert(/Q686F7Z6YU/ =~ cert_data[:certificate].subject.to_s,
           "Wrong data in prod certificate #{cert_data[:certificate].inspect}")
  end
  
  def test_read_push_certificate_on_heroku
    flexmock(Imobile::PushNotifications).
        should_receive(:use_new_certificate_decoder?).and_return(false)
        
    test_read_push_certificate
  end
  
  def test_pack_push_token
    assert_equal @dev_token, Imobile.pack_hex_push_token(@hex_dev_token)
  end
   
  def test_encode_notification
    notification = @notification.merge :push_token => @dev_token
    encoded = Imobile::PushNotifications.encode_notification notification
    
    assert_equal @encoded_notification, encoded
  end
  
  def test_valid_notification
    assert Imobile.valid_notification?(@notification),
           'Failed on easy valid notification'
    notification = @notification.merge :push_token => '1' * 512
    assert Imobile.valid_notification?(notification),
           'Failed on notification with large device token'
    notification[:aps][:alert] = '1' * 512
    assert !Imobile.valid_notification?(notification),
           'Passed notification with large alert'
  end
  
  def test_apns_host_port
    [
     [:push, @dev_cert_path, 'gateway.sandbox.push.apple.com', 2195],
     [:push, @prod_cert_path, 'gateway.push.apple.com', 2195],
     [:feedback, @dev_cert_path, 'feedback.sandbox.push.apple.com', 2196],
     [:feedback, @prod_cert_path, 'feedback.push.apple.com', 2196]
    ].each do |service, cert_path, gold_host, gold_port|
      cert = Imobile::PushNotifications.read_certificate cert_path
      server_type = cert[:server_type]
      assert_equal gold_host,
                   Imobile::PushNotifications.apns_host(server_type, service)
      assert_equal gold_port,
                   Imobile::PushNotifications.apns_port(server_type, service)
    end
  end
    
  def test_smoke_push
    notification = @notification.merge :push_token => @dev_token
    Imobile.push_notification notification, @dev_cert_path
  end
  
  def test_feedback_decoding
    time1 = Time.at((Time.now - 3600).to_i)
    time2 = Time.at((Time.now - 3600 * 72).to_i)
    token1 = @dev_token
    token2 = @dev_token.reverse
    golden_feedback = [
      {:push_token => token1, :time => time1},
      {:push_token => token2, :time => time2}
    ]
    
    reads = [
      [6, [time1.to_i].pack('N')],
      [2, [token1.length].pack('n')],
      [token1.length, token1[0, 10]],
      [token1.length - 10, token1[10, token1.length - 10]],
      [6, [time2.to_i].pack('N')[0, 3]],
      [3, [time2.to_i].pack('N')[3, 1] + [token2.length].pack('n')],
      [token2.length, token2],
      [6, '']
    ]
    
    dev_cert = Imobile::PushNotifications.read_certificate @dev_cert_path
    prod_cert = Imobile::PushNotifications.read_certificate @prod_cert_path
    socket1, socket2 = MockSocket.new(reads), MockSocket.new(reads)
    flexmock(Imobile::PushNotifications).should_receive(:apns_socket).
        with(dev_cert, :feedback).and_return(socket1).once
    flexmock(Imobile::PushNotifications).should_receive(:apns_socket).
        with(prod_cert, :feedback).and_return(socket2)
        
    # Test no-block codepath.
    assert_equal golden_feedback, Imobile.push_feedback(dev_cert),
                 "No-block codepath failed."
    assert socket1.closed, "No-block codepath didn't close socket"
    
    # Test block codepath
    output_feedback = []
    Imobile.push_feedback prod_cert do |feedback_item|
      output_feedback << feedback_item
    end
    assert_equal golden_feedback, output_feedback, "Block codepath failed."
    assert socket2.closed, "Vlock codepath didn't close socket"
  end
  
  def test_smoke_feedback
    Imobile.push_feedback @dev_cert_path do |feedback_item|
      check_feedback_item feedback_item
    end
    
    feedback_items = Imobile.push_feedback @prod_cert_path
    feedback_items.each { |feedback_item| check_feedback_item feedback_item }
  end
  
  # Verifies that a piece of feedback looks valid.
  def check_feedback_item(feedback)
    assert feedback[:push_token],
           "A feedback item does not contain a push token"
    assert feedback[:time],
           "A feedback item does not contain a timestamp"
  end  
end


class MockSocket
  def initialize(expected_reads)
    @expected_reads = expected_reads.dup
    @closed = false
  end
  attr_reader :closed
  
  def read(bytes)
    expected_read = @expected_reads.shift
    unless expected_read
      raise "Unexpected read of #{bytes.inspect} (done sending data)"
    end
    
    unless bytes == expected_read[0]
      raise "Expected read size #{expected_read[0]}, got #{bytes.inspect}"
    end
    expected_read[1]
  end
  
  def close
    raise "Closed prematurely" unless @expected_reads.empty?
    @closed = true
  end
end


lib/imobile/validate_receipt.rb
# Online validation for receipt blobs generated by the In-App Purchase process.
#
# Author:: Victor Costan
# Copyright:: Copyright (C) 2009 Zergling.Net
# License:: MIT

require 'base64'
require 'date'
require 'net/http'
require 'net/https'
require 'openssl'
require 'time'

require 'rubygems'
require 'json'


# :nodoc: namespace
module Imobile

# Decodes and validates an In-App Purchase receipt from the App Store.
#
# Args:
#   receipt_blob:: raw receipt in SKPaymentTransaction.transactionReceipt
#   server_type:: production or sandbox (the API accepts symbols and strings)
#
# The decoded receipt is returned as a Ruby-friendly hash. Keys are converted to
# snake_case symbols (e.g. 'purchase-date' becomes :purchase_date). Dates and
# relevant numbers are parsed out of the JSON strings.
#
# Returns +false+ if validation fails (the receipt was tampered with). Raises a
# RuntimeException if Apple's Web service returns a HTTP error code.
def self.validate_receipt(receipt_blob, server_type = :sandbox)
  AppStoreReceiptValidation.validate_receipt receipt_blob
end


# Implementation details for validate_receipt.
module AppStoreReceiptValidation
  # An URI object pointing to the App Store receipt validating server.
  #
  # The server type is production or sandbox (strings and symbols work).
  def self.store_uri(server_type)
    uri_string = {
      :production => "https://buy.itunes.apple.com/verifyReceipt",
      :sandbox => "https://sandbox.itunes.apple.com/verifyReceipt" 
    }[server_type.to_sym]
    
    uri_string and URI.parse(uri_string)
  end
  
  # A Net:HTTP request for validating a receipt.    
  def self.request(receipt_blob, store_uri)
    request = Net::HTTP::Post.new store_uri.path
    request.set_content_type 'application/json'
    request.body = {'receipt-data' => Base64.encode64(receipt_blob) }.to_json
    request
  end
  
  # Issues a HTTP request to Apple's receipt validating server.
  def self.issue_request(http_request, store_uri)
    http = Net::HTTP.new store_uri.host, store_uri.port
    http.use_ssl = (store_uri.scheme == 'https')
    http.verify_mode = OpenSSL::SSL::VERIFY_NONE
    http.request http_request
  end
  
  # Turns JSON receipt information into a nice Ruby-like receipt.
  #
  # String keys are mapped to Ruby symbols (e.g. 'purchase-date' becomes
  # :purchase_date) and date strings are parsed to Ruby dates.
  def self.nice_receipt(json_receipt)
    nice = Hash[*json_receipt.map { |key, value|
      [key.gsub('-', '_').to_sym, value]
    }.flatten]
    
    [:purchase_date, :original_purchase_date].each do |key|
      nice[key] = DateTime.parse(nice[key]) if nice[key]
    end
    nice[:quantity] = nice[:quantity].to_i if nice[:quantity]
    
    nice
  end
  
  # Processes a HTTP response into a receipt.
  def self.process_response(http_response)
    unless http_response.kind_of? Net::HTTPSuccess
      raise "Internal error in Apple's Web service -- #{http_response.inspect}"
    end
    
    response = JSON.parse http_response.body
    return false unless response['status'] == 0
    nice_receipt response['receipt']
  end
  
  # Real implementation of Imobile.validate_receipt
  def self.validate_receipt(receipt_blob, server_type = :production)
    uri = store_uri server_type
    
    process_response issue_request(request(receipt_blob, uri), uri)     
  end
end  # module AppStoreReceiptValidation

end  # namespace Imobile


lib/imobile/crypto_app_fprint.rb
# Application integrity finger-printing used by ZergSupport's CryptoSupport.  
#
# Author:: Victor Costan
# Copyright:: Copyright (C) 2009 Zergling.Net
# License:: MIT

require 'digest/md5'
require 'digest/sha2'
require 'set'
require 'openssl'


# :nodoc: namespace
module Imobile

# An iPhone application's finger-print, as implemented in CryptoSupport.
#
# Args:
#   device_or_hash:: a Hash or ActiveRecord model representing the result of
#                    calling [ZNDeviceFprint deviceAttributes] on the iMobile
#                    device
#   binary_path:: path to the application's binary (executable file)
#                 corresponding to the application version on the device
#                 (indicated by :app_version in the device attributes)
#
# Returns a finger-print that should prove the application's integrity. The
# finger-print is a string consisting of printable characters.
def self.crypto_app_fprint(device_or_hash, binary_path)
  CryptoSupportAppFprint.app_fprint device_or_hash, binary_path
end


# Implementation details for crypto_app_fprint.
module CryptoSupportAppFprint  
  # The finger-print for a device's attributes, as implemented in CryptoSupport.
  #
  # The device attributes should be passed in a Hash that represents the result
  # of calling [ZNDeviceFprint deviceAttributes] on the iMobile device.
  #  
  # The finger-print is returned as a raw string (no hex-formatting).
  # In particular, the returned finger-print is suitable to be used as a key or
  # IV for AES-128.
  #
  # The code mirrors the reference code in ZergSupport's test suite.
  def self.device_fprint(device_attributes)
    Digest::MD5.digest device_fprint_data(device_attributes)
  end
  
  # The finger-print for a device's attributes, as implemented in CryptoSupport.
  #
  # This method resembles device_fprint, but returns the finger-print in a
  # hex-formatted string, so that it can be used by Web services.  
  def self.hex_device_fprint(device_attributes)
    device_fprint(device_attributes).unpack('C*').map {|c| '%02x' % c}.join('')
  end
  
  # The device data used for device finger-printing in CryptoSupport.
  #
  # The device attributes should be passed in a Hash that represents the result
  # of calling [ZNDeviceFprint deviceAttributes] on the iMobile device.   
  def self.device_fprint_data(device_attributes)
    attrs = device_fprint_attributes
    keys = device_attributes.keys.select { |k| attrs.include? k.to_s }
        'D|' + keys.sort.map { |k| device_attributes[k] }.
        map { |v| v.respond_to?(:read) ? v.read : v }.join('|')
  end

  # The device attributes included in the finger-printing operation.  
  def self.device_fprint_attributes
    Set.new(['app_id', 'app_version', 'app_provisioning', 'app_push_token',
             'hardware_model', 'os_name', 'os_version', 'unique_id'])
  end
  
  # The finger-print for a data blob, as implemented in CryptoSupport.
  #
  # Args:
  #   data_blob:: a raw string, usually the result of reading a file
  #   key:: 16-byte string, to be used as an AES key
  #   iv:: 16-byte string, to be used as an AES key
  #
  # The returned finger-print is a hex-formatted string.
  def self.data_fprint(data_blob, key, iv = "\0" * 16)
    cipher = OpenSSL::Cipher::Cipher.new 'aes-128-cbc'
    cipher.encrypt
    cipher.key, cipher.iv = key, iv
    
    plain = data_blob + "\0" * ((16 - (data_blob.length & 0x0f)) & 0x0f)
    crypted = cipher.update plain
    Digest::SHA2.hexdigest crypted
  end
    
  # An iPhone application's finger-print, as implemented in CryptoSupport.
  #
  # The device attributes should be passed in a Hash that represents the result
  # of calling [ZNDeviceFprint deviceAttributes] on the iMobile device. The
  # manifest data should be the result of reading the application's manifest
  # file (currently its executable file).
  #
  # The returned finger-print is a hex-formatted string.
  def self.app_fprint_from_raw_data(device_attributes, manifest_data)
    key = device_fprint device_attributes
    iv = "\0" * 16
    data_fprint manifest_data, key, iv
  end
  
  # An iPhone application's finger-print, as implemented in CryptoSupport.
  def self.app_fprint(device_or_hash, binary_path)
    if device_or_hash.respond_to?(:[]) and device_or_hash.respond_to?(:keys)
      # Hash-like object.
      device_attributes = device_or_hash
    else
      # ActiveRecord model.
      device_attributes = device_or_hash.attributes
    end
    
    manifest_data = File.read binary_path
    app_fprint_from_raw_data device_attributes, manifest_data
  end    
end  # module CryptoSupportAppFprint

end  # namespace Imobile


lib/imobile/push_notification.rb
# Apple Push Notifications support.
#
# Author:: Victor Costan
# Copyright:: Copyright (C) 2009 Zergling.Net
# License:: MIT

require 'net/http'
require 'net/https'

require 'socket'
require 'openssl'

require 'rubygems'
require 'json'


# :nodoc: namespace
module Imobile

# Sends a push notification to an iMobile device.
#
# Args:
#   notification:: ruby Hash indicating the desired notification; the hash
#                  should have an extra key named :push_token, containing
#                  the binary-encoded (not hexadecimally-encoded) iMobile device
#                  token, as provided by the UIApplicationDelegate method
#                  application:didRegisterForRemoteNotificationsWithDeviceToken: 
#   path_or_certificate:: the certificate required to talk to APNs; this can be
#                         a path to a .p12 file, a string with the contens of
#                         the .p12 file, or a previously-read certificate
#
# Raises a RuntimeException if Apple's Push Notification service doesn't behave.
def self.push_notification(notification, path_or_certificate)
  PushNotifications.push_notification notification, path_or_certificate
end

# Bulk-transmission of push notifications to Apple's service.
#
# Args:
#   path_or_certificate:: see push_notification
#   notifications:: an array of notification hashes; see push_notification
# 
# If the method receives a block, it yields to its block indefinitely. Each
# time, the block should +next+ a notification or array of notifications to be
# pushed. The block should +break+ when it's done.  
def self.push_notifications(path_or_certificate, notifications = [], &block)
  PushNotifications.push_notifications path_or_certificate, notifications,
                                       &block
end

# Reads the available feedback from Apple's Push Notification service.
#
# Args:
#   path_or_certificate:: see Imobile.push_notification
#
# The currently provided feedback is the tokens for the devices which rejected
# notifications. Each piece of feedback is a hash with the following keys:
#   :push_token:: the device's token, in binary (not hexadecimal) format
#   :time:: the last time when the device rejected notifications; according to
#           Apple, the rejection can be discarded if the device sent a
#           token after this time
#
# The method reads all the feedback available from the Push Notification
# service. If a block is given, each piece of feedback is yielded to the
# method's block, and the method returns nil. If no block is given, the 
# method returns an array containing all pieces of feedback.
def self.push_feedback(path_or_certificate, &block)
  PushNotifications.push_feedback path_or_certificate, &block
end

# Checks if a notification is valid for Apple's Push Notification service.
#
# Currently, notifications are valid if their JSON encodings don't exceed 256
# bytes.
def self.valid_notification?(notification)
  PushNotifications.encode_notification(notification) ? true : false
end

# Packs a hexadecimal iMobile token for push notifications into binary form.
def self.pack_hex_push_token(push_token)
  [push_token.gsub(/\s/, '')].pack('H*')
end

# Carries state for delivering batched notifications.
class PushNotificationsContext
  # Sends a notification via this context's APNs connection.
  #
  # Args:
  #   notification:: see Imobile.push_notification
  #
  # Raises a RuntimeError if the context's APNs connection was closed.
  def push(notification)
    raise "The context's APNs connection was closed" if @closed
    @socket.write PushNotifications.encode_notification(notification)
  end
  
  # Creates a push context for a fixed Apple Push Notifications server.
  #
  # Args:
  #   path_or_certificate:: see Imobile.push_notification
  def initialize(path_or_certificate)
    @certificate = PushNotifications.read_certificate path_or_certificate    
    @socket = PushNotifications.apns_socket @certificate, :push
    @closed = false
  end
  
  # Forces a flush of all the buffered data across the APNs connection.
  #
  # This should be called at the end of a batch of notifications. If flush is
  # not called, notifications may be stalled indefinitely.
  def flush
    raise "The context's APNs connection was closed" if @closed
    @socket.flush
  end
  
  # Closes the APNs connection. The context is unusable afterwards.
  def close
    unless @closed
      @socket.flush
      @socket.close
    end      
    @closed = true
  end
  
  # The raw SSL connection to the APNs. 
  attr_reader :socket
  # The APNs client certificate.
  attr_reader :certificate

  # True if the context's APNs connection is closed.
  def closed?
    @closed
  end
  
  # Called when the context is garbage-collected.
  #
  # Closes the APNs connection, if it wasn't already closed.
  def finalize
    close unless @closed
  end
end

# Implementation details for push_notification.
module PushNotifications
  # Reads an APNs certificate from a string or a file.
  def self.read_certificate(certificate_blob_or_path)
    unless certificate_blob_or_path.respond_to? :to_str
      return certificate_blob_or_path
    end    
    begin
      decode_push_certificate File.read(certificate_blob_or_path)
    rescue
      decode_push_certificate certificate_blob_or_path
    end
  end
  
  # Decodes an APNs certificate.
  def self.decode_push_certificate(certificate_blob)
    if use_new_certificate_decoder?
      # Ruby 1.8.7 and above.
      data = decode_push_certificate_new certificate_blob
    else
      # Ruby 1.8.6.
      data = decode_push_certificate_heroku certificate_blob
    end
    data[:server_type] = server_type data[:certificate] 
    data
  end
  
  # Checks whether the new certificate decoding code is supported.
  def self.use_new_certificate_decoder?
    OpenSSL::PKCS12.respond_to? :new    
  end
  
  # Decodes an APNs certificate, using the new (1.8.7+) OpenSSL methods.
  def self.decode_push_certificate_new(certificate_blob)    
    pkcs12 = OpenSSL::PKCS12.new certificate_blob
    
    certificate = pkcs12.certificate
    key = pkcs12.key
    
    { :certificate => certificate, :key => key }
  end
  
  # Decodes an APNs certificate, using the openssl command-line tool.
  #
  # This works on Heroku, which uses Ruby 1.8.6.
  def self.decode_push_certificate_heroku(certificate_blob)
    # Most of the filesystem on Heroku is read-only. On the other hand, not
    # everyone runs on Heroku. Find a reasonable temporary dir.
    if defined? RAILS_ROOT
      temp_dir = File.join RAILS_ROOT, 'tmp'
    elsif File.exists? '/tmp'
      temp_dir = '/tmp'
    else
      temp_dir = '.'
    end
    
    pkcs12_file_name = File.join temp_dir, "apns_#{Process.pid}.p12"
    pem_file_name = File.join temp_dir, "apns_#{Process.pid}.pem"
    out_file_name = File.join temp_dir, "apns_#{Process.pid}.err"
    
    # Use the command-line openssl tool to break up the pkcs12 file.
    File.open(pkcs12_file_name, 'wb') { |f| f.write certificate_blob }
    Kernel.system "openssl pkcs12 -in #{pkcs12_file_name} -clcerts -nodes " +
                  "-out #{pem_file_name} -password pass: 2> #{out_file_name}"
    pem_blob = File.read pem_file_name    
    [pkcs12_file_name, pem_file_name, out_file_name].each { |f| File.delete f }
    
    certificate = OpenSSL::X509::Certificate.new pem_blob
    key = OpenSSL::PKey::RSA.new pem_blob
    { :certificate => certificate, :key => key }    
  end
  
  # The Apple Push Notification server type that a certificate works with.
  def self.server_type(certificate)
    case certificate.subject.to_s
    when /Apple Development Push/
      return :sandbox
    when /Apple Production Push/
      return :production
    else
      raise "Invalid push certificate - #{certificate.inspect}"
    end        
  end
  
  # Encodes a push notification in a binary string for APNs consumption.
  #
  # Returns a string suitable for transmission over an APNs, or nil if the
  # notification is invalid (i.e. the json encoding exceeds 256 bytes).
  def self.encode_notification(notification)
    push_token = notification[:push_token] || ''
    notification = notification.dup
    notification.delete :push_token
    json_notification = notification.to_json
    return nil if json_notification.length > 256
    
    ["\0", [push_token.length].pack('n'), push_token,
     [json_notification.length].pack('n'), json_notification].join
  end
    
  # Creates a socket to an Apple Push Notification Server.
  #
  # Args:
  #   push_certificate:: the APNs client certificate data, obtained by a call to
  #                      read_certificate
  #   service:: either :feedback or :push
  #    
  # The returned socket is connected and ready for use.
  def self.apns_socket(push_certificate, service = :push)
    context = OpenSSL::SSL::SSLContext.new
    context.cert = push_certificate[:certificate]
    context.key = push_certificate[:key]
    
    server_type = push_certificate[:server_type]
    raw_socket = TCPSocket.new apns_host(server_type, service),
                               apns_port(server_type, service)
    
    socket = OpenSSL::SSL::SSLSocket.new raw_socket, context
    # Magic for closing the raw socket when the SSL socket is closed.
    (class <<socket; self; end).send :define_method, :close do
      super
      raw_socket.close
    end
    socket.connect
  end
  
  # The host name for an Apple Push Notification Server.
  #
  # Args:
  #   server_type:: either :production or :sandbox
  #   service:: either :push or :feedback
  def self.apns_host(server_type, service = :push)
    {
      :feedback => {
        :sandbox => 'feedback.sandbox.push.apple.com',
        :production => 'feedback.push.apple.com'
      },
      :push => {
        :sandbox => 'gateway.sandbox.push.apple.com',
        :production => 'gateway.push.apple.com'
      }
    }[service][server_type]
  end
  
  # The port for an Apple Push Notification Server.
  #
  # Args:
  #   server_type:: either :production or :sandbox
  #   service:: either :push or :feedback
  def self.apns_port(server_type, service = :push)
    {
      :feedback => 2196,
      :push => 2195
    }[service]
  end
  
  # Real implementation of Imobile.push_notifications
  def self.push_notifications(certificate_or_path, notifications)
    context = PushNotificationsContext.new certificate_or_path
    notifications = [notifications] if notifications.kind_of? Hash
    notifications.each { |notification| context.push notification }
    if Kernel.block_given?
      loop do
        notifications = yield
        notifications = [notifications] if notifications.kind_of? Hash
        notifications.each { |notification| context.push notification }
      end
    end
    context.flush
    context.close
  end
  
  # Real implementation of Imobile.push_notification
  def self.push_notification(notification, certificate_or_path)
    push_notifications certificate_or_path, [notification]
  end

  # Real implementation of Imobile.push_feedback
  def self.push_feedback(certificate_or_path, &block)
    if Kernel.block_given?
      raw_push_feedback certificate_or_path, &block
      nil
    else
      feedback = []
      raw_push_feedback certificate_or_path do |feedback_item|
        feedback << feedback_item
      end
      feedback
    end
  end
  
  # Reads the available feedback from Apple's Push Notification service.
  #
  # Args:
  #   certificate_or_path:: see Imobile.push_notification
  #
  # The currently provided feedback is the tokens for the devices which rejected
  # notifications. Each piece of feedback is a hash with the following keys:
  #   :push_token:: the device's token for push notifications, in binary
  #                 (not hexadecimal) format
  #   :time:: the last time when the device rejected notifications; according to
  #           Apple, the rejection can be discarded if the device sent a
  #           token after this time
  #
  # The method reads all the feedback available from the Push Notification
  # service, and yields each piece of feedback to the method's block.
  def self.raw_push_feedback(certificate_or_path)
    socket = apns_socket read_certificate(certificate_or_path), :feedback
    loop do      
      break unless header = fixed_socket_read(socket, 6)
      time = Time.at header[0, 4].unpack('N').first
      push_token = fixed_socket_read(socket, header[4, 2].unpack('n').first)
      break unless push_token
      feedback_item = { :push_token => push_token, :time => time }
      yield feedback_item
    end
    socket.close
  end
  
  # Reads a fixed number of bytes from a socket.
  def self.fixed_socket_read(socket, num_bytes)
    data = ''
    while data.length < num_bytes
      new_data = socket.read(num_bytes - data.length)
      return nil if new_data.nil? or new_data.empty?  # Socket closed.
      data += new_data
    end
    data
  end
end  # module PushNotifications

end  # namespace Imobile
